diff --git a/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/java/org/apache/hadoop/fs/http/server/FSOperations.java b/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/java/org/apache/hadoop/fs/http/server/FSOperations.java
index 11cdb4d02c..57bf025f80 100644
--- a/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/java/org/apache/hadoop/fs/http/server/FSOperations.java
+++ b/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/java/org/apache/hadoop/fs/http/server/FSOperations.java
@@ -1025,7 +1025,7 @@ public static class FSRemoveAclEntries implements FileSystemAccess.FileSystemExe
      */
     public FSRemoveAclEntries(String path, String aclSpec) {
       this.path = new Path(path);
-      this.aclEntries = AclEntry.parseAclSpec(aclSpec, true);
+      this.aclEntries = AclEntry.parseAclSpec(aclSpec, false);
     }
 
     /**
